GetIntoTeachingApi/Services/Store.cs

```diff
@@ -145,21 +151,22 @@ private async Task<IEnumerable<TeachingEvent>> OnlineEventsMatchingRequest(Teach
             return result.Where(te => te.IsOnline && !te.IsVirtual);
         }
 
+        private async Task SyncTeachingEventBuildings()
+        {
+            var buildings = _crm.GetTeachingEventBuildings();
+            await PopulateTeachingEventCoordinates(buildings);
+
+            await SyncModels(buildings, _dbContext.TeachingEventBuildings);
+        }
+
         private async Task SyncTeachingEvents()
         {
             var afterDate = DateTime.UtcNow.Subtract(TeachingEventArchiveSize);
             var teachingEvents = _crm.GetTeachingEvents(afterDate).ToList();
-            await PopulateTeachingEventCoordinates(teachingEvents);
-
-            var buildings = teachingEvents.Where(te => te.Building != null)
-                .Select(te => te.Building).DistinctBy(b => b.Id);
-
-            await SyncModels(buildings, _dbContext.TeachingEventBuildings);
 
-            // Link events with buildings attached to the context prior to sync.
             foreach (var te in teachingEvents.Where(te => te.Building != null))
             {
-                te.Building = await _dbContext.TeachingEventBuildings.FindAsync(te.Building.Id);
+                te.Building = await _dbContext.TeachingEventBuildings.FindAsync(te.BuildingId);

```

Comment:
It may be more efficient to load all buildings into memory and run `Find` on that instead of hitting the database for each teaching event